File: packages/updater-core/ios/Sources/LaunchStore.swift
import Darwin
import Foundation

public struct Artifact: Codable, Equatable {
  public let appId: String
  public let platform: String
  public let runtimeVersion: String
  public let contentHash: String
  public let version: String
  public let releaseId: String?
  public let channel: String?
  public let bundlePath: String
  public let embedded: Bool

  public init(
    appId: String, platform: String, runtimeVersion: String, contentHash: String,
    version: String, releaseId: String? = nil, channel: String? = nil, bundlePath: String,
    embedded: Bool = false
  ) {
    self.appId = appId
    self.platform = platform
    self.runtimeVersion = runtimeVersion
    self.contentHash = contentHash
    self.version = version
    self.releaseId = releaseId
    self.channel = channel
    self.bundlePath = bundlePath
    self.embedded = embedded
  }
}

public struct TrialEvent: Codable, Equatable {
  public let id: String
  public let type: String
  public let artifact: Artifact
  public var createdAt: TimeInterval? = Date().timeIntervalSince1970
  public var detail: String? = nil
}

public struct LaunchSnapshot: Codable {
  public var schemaVersion = 1
  public var buildId: String
  public var generation: UInt64 = 0
  public var current: Artifact
  public var lastGood: Artifact
  public var previousGood: Artifact?
  public var staged: Artifact?
  public var trialGeneration: UInt64?
  public var failed: Set<String> = []
  public var events: [TrialEvent] = []
  public var droppedEvents: Int?
  public var lastEventError: String?
}

extension LaunchSnapshot {
  public init(from decoder: Decoder) throws {
    let fields = try decoder.container(keyedBy: CodingKeys.self)
    self.init(
      buildId: try fields.decode(String.self, forKey: .buildId),
      current: try fields.decode(Artifact.self, forKey: .current),
      lastGood: try fields.decode(Artifact.self, forKey: .lastGood))
    schemaVersion = try fields.decode(Int.self, forKey: .schemaVersion)
    generation = try fields.decode(UInt64.self, forKey: .generation)
    previousGood = try fields.decodeIfPresent(Artifact.self, forKey: .previousGood)
    staged = try fields.decodeIfPresent(Artifact.self, forKey: .staged)
    trialGeneration = try fields.decodeIfPresent(UInt64.self, forKey: .trialGeneration)
    failed = try fields.decode(Set<String>.self, forKey: .failed)
    droppedEvents = try? fields.decodeIfPresent(Int.self, forKey: .droppedEvents)
    lastEventError = try? fields.decodeIfPresent(String.self, forKey: .lastEventError)
    // A malformed telemetry entry must not discard otherwise valid launch/recovery pointers.
    var invalid = 0
    if var outbox = try? fields.nestedUnkeyedContainer(forKey: .events) {
      while !outbox.isAtEnd {
        let item = try outbox.superDecoder()
        if let event = try? TrialEvent(from: item) { events.append(event) } else { invalid += 1 }
      }
    } else {
      invalid = 1
    }
    if invalid > 0 {
      droppedEvents = max(0, droppedEvents ?? 0) + invalid
      lastEventError = "Invalid saved event payload"
    }
  }
}

public enum LaunchError: Error {
  case incompatibleArtifact, quarantined, activationDeferred, staleInstance, storageUnavailable,
    corruptState
}

/// One application-lifetime owner. Module bindings capture a generation; they never own this store.
public final class LaunchStore {
  private let queue = DispatchQueue(label: "com.otakit.rn.launch-store")
  private let file: URL
  private let builtin: Artifact
  private var snapshot: LaunchSnapshot
  private var bootstrapGeneration: UInt64?
  private var hostGeneration: UInt64?
  private var didBeginLaunch = false
  private var storageFailed = false
  private var foreground = false
  private var guardActive = false
  private var backgroundWorkObserved = false
  private var trialElapsed: TimeInterval = 0
  private var foregroundSince: TimeInterval?
  private let readyTimeout: TimeInterval

  public init(
    file: URL, buildId: String, builtin: Artifact, readyTimeout: TimeInterval = 10,
    validateCachedArtifact: ((Artifact) throws -> Bool)? = nil
  ) throws {
    self.file = file
    self.builtin = builtin
    self.readyTimeout = readyTimeout
    let initial = LaunchSnapshot(buildId: buildId, current: builtin, lastGood: builtin)
    do {
      let data = try Data(contentsOf: file)
      let stored = try JSONDecoder().decode(LaunchSnapshot.self, from: data)
      if stored.schemaVersion == 1 && stored.generation < UInt64.max - 1
        && stored.buildId == buildId && Self.compatible(stored.current, builtin)
        && Self.compatible(stored.lastGood, builtin)
      {
        snapshot = stored
      } else {
        snapshot = initial
      }
    } catch let error as CocoaError where error.code == .fileReadNoSuchFile {
      snapshot = initial
    } catch let error as CocoaError where error.code == .fileReadNoPermission {
      throw LaunchError.storageUnavailable
    } catch is DecodingError { snapshot = initial } catch { throw LaunchError.storageUnavailable }
    // A prior process died before readiness. Resolve once before selecting any code.
    if snapshot.trialGeneration != nil {
      var recovered = snapshot
      Self.failTrial(&recovered)
      try persist(recovered)
    }
    if let validate = validateCachedArtifact {
      var checked = snapshot
      if !(try validate(checked.lastGood)) { checked.lastGood = builtin }
      if !(try validate(checked.current)) { checked.current = checked.lastGood }
      if let staged = checked.staged, !(try validate(staged)) { checked.staged = nil }
      if let previous = checked.previousGood, !(try validate(previous)) {
        checked.previousGood = nil
      }
      // A verified embedded identity always uses the resource path of this installed binary.
      if checked.current.embedded {
        checked.current = Self.relocateEmbedded(checked.current, builtin)
      }
      if checked.lastGood.embedded {
        checked.lastGood = Self.relocateEmbedded(checked.lastGood, builtin)
      }
      if let staged = checked.staged, staged.embedded {
        checked.staged = Self.relocateEmbedded(staged, builtin)
      }
      try persist(checked)
    }
  }

  public func state() -> LaunchSnapshot { queue.sync { snapshot } }

  public func stage(_ artifact: Artifact, from generation: UInt64? = nil, downloaded: Bool = false)
    throws
  {
    try queue.sync {
      if let generation = generation {
        guard generation == snapshot.generation && bootstrapGeneration == generation else {
          throw LaunchError.staleInstance
        }
      }
      guard Self.compatible(artifact, builtin) else { throw LaunchError.incompatibleArtifact }
      guard !snapshot.failed.contains(artifact.contentHash) else { throw LaunchError.quarantined }
      var next = snapshot
      if artifact.contentHash == next.current.contentHash {
        // Association changes never confirm or reattribute an executing trial.
        guard next.trialGeneration == nil else { return }
        next.current = artifact
        Self.confirm(artifact, in: &next)
        next.staged = nil
        try persist(next)
        return
      }
      next.staged = artifact
      if downloaded {
        next.events.append(
          TrialEvent(id: UUID().uuidString, type: "downloaded", artifact: artifact))
      }
      try persist(next)
    }
  }

  /// Cold startup calls once immediately before handing a loader to RN; URL callbacks only read state().
  public func beginLaunch(foreground: Bool, activateStaged: Bool, now: TimeInterval) throws
    -> UInt64
  {
    try queue.sync {
      if didBeginLaunch { return snapshot.generation }
      self.foreground = foreground
      if !foreground { backgroundWorkObserved = true }
      let generation = try select(
        activateStaged: foreground && !guardActive && activateStaged, now: now)
      didBeginLaunch = true
      return generation
    }
  }

  public func bindBootstrap(generation: UInt64) throws {
    try queue.sync {
      guard generation == snapshot.generation else { throw LaunchError.staleInstance }
      bootstrapGeneration = generation
    }
  }

  public func assertInstance(_ generation: UInt64) throws {
    try queue.sync {
      guard generation == snapshot.generation && bootstrapGeneration == generation else {
        throw LaunchError.staleInstance
      }
    }
  }

  public func hostReady(generation: UInt64) {
    queue.sync { if generation == snapshot.generation { hostGeneration = generation } }
  }

  public func setForeground(_ value: Bool, now: TimeInterval) {
    queue.sync {
      accountTime(now)
      foreground = value
      foregroundSince = value && snapshot.trialGeneration != nil ? now : nil
    }
  }

  public func setActivationGuard(_ value: Bool, from generation: UInt64? = nil) {
    queue.sync {
      if let generation = generation,
        generation != snapshot.generation || bootstrapGeneration != generation
      {
        return
      }
      guardActive = value
    }
  }

  public func observeBackgroundWork(from generation: UInt64) {
    queue.sync {
      if generation == snapshot.generation { backgroundWorkObserved = true }
    }
  }

  public func apply(from generation: UInt64, now: TimeInterval) throws -> UInt64 {
    try queue.sync {
      guard generation == snapshot.generation && bootstrapGeneration == generation else {
        throw LaunchError.staleInstance
      }
      guard
        foreground && !guardActive && !backgroundWorkObserved && hostGeneration == generation
          && snapshot.trialGeneration == nil
      else { throw LaunchError.activationDeferred }
      guard snapshot.staged != nil else { return generation }
      return try select(activateStaged: true, now: now)
    }
  }

  public func notifyReady(from generation: UInt64) throws {
    try queue.sync {
      guard generation == snapshot.generation && bootstrapGeneration == generation else {
        throw LaunchError.staleInstance
      }
      guard snapshot.trialGeneration == generation else { return }
      var next = snapshot
      Self.confirm(next.current, in: &next)
      next.trialGeneration = nil
      next.events.append(TrialEvent(id: UUID().uuidString, type: "applied", artifact: next.current))
      try persist(next)
      foregroundSince = nil
      trialElapsed = 0
    }
  }

  public func checkTimeout(now: TimeInterval) throws -> Bool {
    try queue.sync {
      accountTime(now)
      guard snapshot.trialGeneration != nil && trialElapsed >= readyTimeout else { return false }
      return try failCurrent()
    }
  }

  public func fail(from generation: UInt64) throws -> Bool {
    try queue.sync {
      guard generation == snapshot.trialGeneration else { return false }
      return try failCurrent()
    }
  }

  public func acknowledgeEvents(_ ids: Set<String>) throws {
    try queue.sync {
      var next = snapshot
      next.events.removeAll { ids.contains($0.id) }
      try persist(next)
    }
  }

  public func recordDownloadFailure(_ artifact: Artifact, detail: String) throws {
    try queue.sync {
      guard Self.compatible(artifact, builtin), artifact.releaseId != nil else { return }
      var next = snapshot
      var event = TrialEvent(id: UUID().uuidString, type: "download_error", artifact: artifact)
      event.detail = String(detail.prefix(500))
      next.events.append(event)
      try persist(next)
    }
  }

  public func eventForDelivery() throws -> TrialEvent? {
    try queue.sync {
      var next = snapshot
      Self.trimEvents(&next)
      let changed = next.events.count != snapshot.events.count
      let needsTimestamp = next.events.first != nil && next.events[0].createdAt == nil
      if needsTimestamp { next.events[0].createdAt = Date().timeIntervalSince1970 }
      if changed || needsTimestamp { try persist(next) }
      return snapshot.events.first
    }
  }

  public func discardEvent(_ id: String, reason: String) throws {
    try queue.sync {
      guard snapshot.events.contains(where: { $0.id == id }) else { return }
      var next = snapshot
      next.events.removeAll { $0.id == id }
      next.droppedEvents = (next.droppedEvents ?? 0) + 1
      next.lastEventError = String(reason.prefix(500))
      try persist(next)
    }
  }

  private static func trimEvents(_ state: inout LaunchSnapshot) {
    let count = state.events.count
    let now = Date().timeIntervalSince1970
    state.events.removeAll { $0.createdAt.map { now - $0 > 86400 } ?? false }
    if state.events.count > 256 { state.events.removeFirst(state.events.count - 256) }
    if state.events.count != count {
      state.droppedEvents = (state.droppedEvents ?? 0) + count - state.events.count
      state.lastEventError = "Event retention limit exceeded"
    }
  }

  private func select(activateStaged: Bool, now: TimeInterval) throws -> UInt64 {
    var next = snapshot
    // No second launch may erase an unresolved trial.
    if next.trialGeneration != nil { Self.failTrial(&next) }
    next.generation += 1
    if activateStaged, let staged = next.staged {
      guard Self.compatible(staged, builtin) else { throw LaunchError.incompatibleArtifact }
      guard !next.failed.contains(staged.contentHash) else { throw LaunchError.quarantined }
      next.current = staged
      next.staged = nil
      // Switching back to embedded code still needs readiness from the new instance.
      if staged.contentHash != next.lastGood.contentHash {
        next.trialGeneration = next.generation
      } else {
        Self.confirm(staged, in: &next)
      }
    } else {
      next.current = next.lastGood
    }
    try persist(next)
    bootstrapGeneration = nil
    hostGeneration = nil
    trialElapsed = 0
    guardActive = false
    foregroundSince = foreground && next.trialGeneration != nil ? now : nil
    return next.generation
  }

  private func accountTime(_ now: TimeInterval) {
    if let since = foregroundSince {
      trialElapsed += max(0, now - since)
      foregroundSince = now
    }
  }

  private func failCurrent() throws -> Bool {
    var next = snapshot
    Self.failTrial(&next)
    next.generation += 1
    try persist(next)
    bootstrapGeneration = nil
    hostGeneration = nil
    foregroundSince = nil
    guardActive = false
    return true
  }

  private static func failTrial(_ state: inout LaunchSnapshot) {
    guard state.trialGeneration != nil else { return }
    state.failed.insert(state.current.contentHash)
    state.events.append(
      TrialEvent(id: UUID().uuidString, type: "rollback", artifact: state.current))
    state.current = state.lastGood
    state.trialGeneration = nil
  }

  private static func confirm(_ artifact: Artifact, in state: inout LaunchSnapshot) {
    if state.lastGood.contentHash != artifact.contentHash && !state.lastGood.embedded {
      state.previousGood = state.lastGood
    }
    state.lastGood = artifact
  }

  private static func compatible(_ artifact: Artifact, _ builtin: Artifact) -> Bool {
    artifact.appId == builtin.appId && artifact.platform == builtin.platform
      && artifact.runtimeVersion == builtin.runtimeVersion
  }

  private static func relocateEmbedded(_ artifact: Artifact, _ builtin: Artifact) -> Artifact {
    Artifact(
      appId: builtin.appId, platform: builtin.platform, runtimeVersion: builtin.runtimeVersion,
      contentHash: builtin.contentHash, version: artifact.version, releaseId: artifact.releaseId,
      channel: artifact.channel, bundlePath: builtin.bundlePath, embedded: true)
  }

  private func persist(_ next: LaunchSnapshot) throws {
    var next = next
    Self.trimEvents(&next)
    guard !storageFailed else { throw LaunchError.storageUnavailable }
    // Any failed write can have an uncertain durable outcome. Stop this owner until cold recovery.
    storageFailed = true
    let directory = file.deletingLastPathComponent()
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let temporary = directory.appendingPathComponent(".\(UUID().uuidString).tmp")
    defer { try? FileManager.default.removeItem(at: temporary) }
    let data = try JSONEncoder().encode(next)
    try data.write(to: temporary, options: [.withoutOverwriting])
    let fd = Darwin.open(temporary.path, O_RDONLY)
    guard fd >= 0 else { throw LaunchError.storageUnavailable }
    let synced = fsync(fd)
    close(fd)
    guard synced == 0 && rename(temporary.path, file.path) == 0 else {
      throw LaunchError.storageUnavailable
    }
    let dirfd = Darwin.open(directory.path, O_RDONLY)
    guard dirfd >= 0 else { throw LaunchError.storageUnavailable }
    let dirSynced = fsync(dirfd)
    close(dirfd)
    guard dirSynced == 0 else { throw LaunchError.storageUnavailable }
    snapshot = next
    storageFailed = false
  }
}

File: packages/updater-core/android/src/main/java/com/otakit/core/LaunchStore.java
package com.otakit.core;

import java.io.File;
import java.io.FileOutputStream;
import java.nio.channels.FileChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.UUID;
import org.json.JSONArray;
import org.json.JSONObject;

/** Application-lifetime serialized launch state. RN module bindings retain one generation. */
public final class LaunchStore {

  private final File file;
  private final JSONObject builtin;
  private final DirectorySync directorySync;

  public interface DirectorySync {
    void sync(File directory) throws Exception;
  }

  public interface ArtifactVerifier {
    boolean verify(JSONObject artifact) throws Exception;
  }

  private JSONObject state;
  private boolean foreground, guard, began, storageFailed;
  private boolean backgroundWorkObserved;
  private long bootstrap = -1,
    host = -1;
  private long elapsedMillis,
    foregroundSince = -1;

  public LaunchStore(File file, String buildId, JSONObject builtin) throws Exception {
    this(
      file,
      buildId,
      builtin,
      directory -> {
        try (FileChannel channel = FileChannel.open(directory.toPath(), StandardOpenOption.READ)) {
          channel.force(true);
        }
      },
      artifact -> true
    );
  }

  public LaunchStore(
    File file,
    String buildId,
    JSONObject builtin,
    DirectorySync directorySync,
    ArtifactVerifier verifier
  ) throws Exception {
    this.file = file;
    this.builtin = copy(builtin);
    this.directorySync = directorySync;
    state = new JSONObject()
      .put("schemaVersion", 1)
      .put("buildId", buildId)
      .put("generation", 0)
      .put("current", copy(builtin))
      .put("lastGood", copy(builtin))
      .put("failed", new JSONArray())
      .put("events", new JSONArray());
    if (file.exists()) {
      // An I/O error (including protected/unavailable storage) must not erase state.
      String saved = new String(Files.readAllBytes(file.toPath()), StandardCharsets.UTF_8);
      try {
        JSONObject parsed = new JSONObject(saved);
        if (
          parsed.getInt("schemaVersion") == 1 &&
          RNManifest.integer(parsed, "generation") >= 0 &&
          RNManifest.integer(parsed, "generation") < 9007199254740990L &&
          buildId.equals(parsed.getString("buildId")) &&
          compatible(parsed.getJSONObject("current")) &&
          compatible(parsed.getJSONObject("lastGood"))
        ) {
          parsed.getJSONArray("failed");
          parsed.getJSONArray("events");
          state = parsed;
        }
      } catch (Exception error) {
        /* Corrupt state selects the embedded application. */
      }
    }
    JSONObject checked = copy(state);
    for (String key : new String[] { "current", "lastGood", "staged", "previousGood" }) {
      if (!checked.has(key)) continue;
      JSONObject artifact = checked.optJSONObject(key);
      boolean valid = false;
      try {
        valid = artifact != null && verifier.verify(artifact);
      } catch (org.json.JSONException error) {
        /* Invalid pointer, not unavailable storage. */
      }
      if (!valid) {
        if (key.equals("staged") || key.equals("previousGood")) checked.remove(key);
        else checked.put(key, copy(builtin));
      } else if (artifact.optBoolean("embedded")) artifact.put(
        "bundlePath",
        builtin.getString("bundlePath")
      );
    }
    // Keep the original failed trial identity for quarantine and attribution.
    if (state.has("trialGeneration")) checked.put("current", copy(state.getJSONObject("current")));
    state = checked;
    if (state.has("trialGeneration")) {
      JSONObject next = copy(state);
      failTrial(next);
      persist(next);
    } else {
      // Publish verified pointers durably before background collection can remove rejected code.
      persist(checked);
    }
  }

  public synchronized JSONObject state() throws Exception {
    return copy(state);
  }

  public synchronized void stage(JSONObject artifact) throws Exception {
    stageArtifact(artifact, false);
  }

  private void stageArtifact(JSONObject artifact, boolean downloaded) throws Exception {
    if (!compatible(artifact)) throw new Exception("INCOMPATIBLE_ARTIFACT");
    if (failed(artifact.getString("contentHash"))) throw new Exception("QUARANTINED_ARTIFACT");
    JSONObject next = copy(state);
    if (
      artifact
        .getString("contentHash")
        .equals(next.getJSONObject("current").getString("contentHash"))
    ) {
      if (next.has("trialGeneration")) return;
      next.put("current", copy(artifact));
      confirm(next, artifact);
      next.remove("staged");
      persist(next);
      return;
    }
    next.put("staged", copy(artifact));
    if (downloaded) event(next, "downloaded", artifact, null);
    persist(next);
  }

  public synchronized void assertInstance(long generation) throws Exception {
    if (generation != state.getLong("generation") || bootstrap != generation) throw new Exception(
      "STALE_INSTANCE"
    );
  }

  public synchronized void stage(JSONObject artifact, long generation) throws Exception {
    assertInstance(generation);
    stage(artifact);
  }

  public synchronized void stage(JSONObject artifact, long generation, boolean downloaded)
    throws Exception {
    assertInstance(generation);
    stageArtifact(artifact, downloaded);
  }

  public synchronized void recordDownloadFailure(JSONObject artifact, String detail)
    throws Exception {
    if (!compatible(artifact) || artifact.isNull("releaseId")) return;
    JSONObject next = copy(state);
    event(next, "download_error", artifact, detail);
    persist(next);
  }

  public synchronized JSONObject eventForDelivery() throws Exception {
    JSONObject next = copy(state);
    trimEvents(next);
    JSONArray events = next.getJSONArray("events");
    boolean changed = events.length() != state.getJSONArray("events").length();
    if (events.length() > 0 && events.getJSONObject(0).isNull("createdAt")) {
      events.getJSONObject(0).put("createdAt", System.currentTimeMillis() / 1000.0);
      changed = true;
    }
    if (changed) persist(next);
    return events.length() == 0 ? null : copy(events.getJSONObject(0));
  }

  public synchronized void discardEvent(String id, String reason) throws Exception {
    JSONObject next = copy(state);
    JSONArray events = next.getJSONArray("events");
    for (int index = 0; index < events.length(); index++) {
      if (events.getJSONObject(index).getString("id").equals(id)) {
        events.remove(index);
        next.put("droppedEvents", next.optLong("droppedEvents") + 1);
        next.put("lastEventError", reason.substring(0, Math.min(500, reason.length())));
        persist(next);
        return;
      }
    }
  }

  private static void trimEvents(JSONObject state) throws Exception {
    JSONArray before = state.getJSONArray("events");
    JSONArray kept = new JSONArray();
    double now = System.currentTimeMillis() / 1000.0;
    for (int index = 0; index < before.length(); index++) {
      JSONObject event = before.optJSONObject(index);
      if (event == null || event.optString("id").isEmpty()) continue;
      double created = event.optDouble("createdAt", Double.NaN);
      if (
        event.isNull("createdAt") || (Double.isFinite(created) && now - created <= 86400)
      ) kept.put(event);
    }
    while (kept.length() > 256) kept.remove(0);
    if (kept.length() != before.length()) {
      state.put("events", kept);
      state.put("droppedEvents", state.optLong("droppedEvents") + before.length() - kept.length());
      state.put("lastEventError", "Invalid event or retention limit exceeded");
    }
  }

  public synchronized long beginLaunch(
    boolean foreground,
    boolean activateStaged,
    long monotonicMillis
  ) throws Exception {
    if (began) return state.getLong("generation");
    this.foreground = foreground;
    if (!foreground) backgroundWorkObserved = true;
    long generation = select(foreground && !guard && activateStaged, monotonicMillis);
    began = true;
    return generation;
  }

  public synchronized void bindBootstrap(long generation) throws Exception {
    bindLaunch(generation);
  }

  /** Capture the validated launch before recovery can advance to another artifact. */
  public synchronized JSONObject bindLaunch(long generation) throws Exception {
    if (generation != state.getLong("generation")) throw new Exception("STALE_INSTANCE");
    bootstrap = generation;
    return copy(state);
  }

  public synchronized void hostReady(long generation) throws Exception {
    if (generation == state.getLong("generation")) host = generation;
  }

  public synchronized void setForeground(boolean value, long now) {
    account(now);
    foreground = value;
    foregroundSince = value && state.has("trialGeneration") ? now : -1;
  }

  public synchronized void setActivationGuard(boolean active) {
    guard = active;
  }

  public synchronized void observeBackgroundWork(long generation) throws Exception {
    if (generation == state.getLong("generation")) backgroundWorkObserved = true;
  }

  public synchronized void setActivationGuard(boolean active, long generation) throws Exception {
    if (generation == state.getLong("generation") && bootstrap == generation) guard = active;
  }

  public synchronized long apply(long generation, long now) throws Exception {
    if (generation != state.getLong("generation") || bootstrap != generation) throw new Exception(
      "STALE_INSTANCE"
    );
    if (
      !foreground ||
      guard ||
      backgroundWorkObserved ||
      host != generation ||
      state.has("trialGeneration")
    ) throw new Exception("ACTIVATION_DEFERRED");
    if (!state.has("staged")) return generation;
    return select(true, now);
  }

  public synchronized void notifyReady(long generation) throws Exception {
    if (generation != state.getLong("generation") || bootstrap != generation) throw new Exception(
      "STALE_INSTANCE"
    );
    if (!state.has("trialGeneration") || state.getLong("trialGeneration") != generation) return;
    JSONObject next = copy(state);
    confirm(next, next.getJSONObject("current"));
    next.remove("trialGeneration");
    event(next, "applied");
    persist(next);
    foregroundSince = -1;
    elapsedMillis = 0;
  }

  public synchronized boolean checkTimeout(long now, long timeoutMillis) throws Exception {
    account(now);
    return state.has("trialGeneration") && elapsedMillis >= timeoutMillis && failCurrent();
  }

  public synchronized boolean fail(long generation) throws Exception {
    return (
      state.has("trialGeneration") &&
      generation == state.getLong("trialGeneration") &&
      failCurrent()
    );
  }

  public synchronized void acknowledgeEvents(java.util.Set<String> ids) throws Exception {
    JSONObject next = copy(state);
    JSONArray kept = new JSONArray();
    JSONArray events = next.getJSONArray("events");
    for (int i = 0; i < events.length(); i++) if (
      !ids.contains(events.getJSONObject(i).getString("id"))
    ) kept.put(events.getJSONObject(i));
    next.put("events", kept);
    persist(next);
  }

  private long select(boolean activate, long now) throws Exception {
    JSONObject next = copy(state);
    if (next.has("trialGeneration")) failTrial(next);
    long generation = next.getLong("generation") + 1;
    next.put("generation", generation);
    if (activate && next.has("staged")) {
      JSONObject staged = next.getJSONObject("staged");
      if (!compatible(staged) || failed(staged.getString("contentHash"))) throw new Exception(
        "INCOMPATIBLE_ARTIFACT"
      );
      next.put("current", staged);
      next.remove("staged");
      String content = staged.getString("contentHash");
      if (!content.equals(next.getJSONObject("lastGood").getString("contentHash"))) next.put(
        "trialGeneration",
        generation
      );
      else confirm(next, staged);
    } else next.put("current", copy(next.getJSONObject("lastGood")));
    persist(next);
    bootstrap = -1;
    host = -1;
    guard = false;
    elapsedMillis = 0;
    foregroundSince = foreground && next.has("trialGeneration") ? now : -1;
    return generation;
  }

  private boolean failCurrent() throws Exception {
    JSONObject next = copy(state);
    failTrial(next);
    next.put("generation", next.getLong("generation") + 1);
    persist(next);
    bootstrap = -1;
    host = -1;
    guard = false;
    foregroundSince = -1;
    return true;
  }

  private static void failTrial(JSONObject state) throws Exception {
    if (!state.has("trialGeneration")) return;
    state.getJSONArray("failed").put(state.getJSONObject("current").getString("contentHash"));
    event(state, "rollback");
    state.put("current", copy(state.getJSONObject("lastGood")));
    state.remove("trialGeneration");
  }

  private static void confirm(JSONObject state, JSONObject artifact) throws Exception {
    JSONObject lastGood = state.getJSONObject("lastGood");
    if (
      !lastGood.getString("contentHash").equals(artifact.getString("contentHash")) &&
      !lastGood.optBoolean("embedded")
    ) state.put("previousGood", copy(lastGood));
    state.put("lastGood", copy(artifact));
  }

  private static void event(JSONObject state, String type) throws Exception {
    event(state, type, state.getJSONObject("current"), null);
  }

  private static void event(JSONObject state, String type, JSONObject artifact, String detail)
    throws Exception {
    state
      .getJSONArray("events")
      .put(
        new JSONObject()
          .put("id", UUID.randomUUID().toString())
          .put("type", type)
          .put("artifact", copy(artifact))
          .put("createdAt", System.currentTimeMillis() / 1000.0)
          .put(
            "detail",
            detail == null ? JSONObject.NULL : detail.substring(0, Math.min(500, detail.length()))
          )
      );
  }

  private void account(long now) {
    if (foregroundSince >= 0) {
      elapsedMillis += Math.max(0, now - foregroundSince);
      foregroundSince = now;
    }
  }

  private boolean compatible(JSONObject artifact) throws Exception {
    // Recovery needs the original trial hash even when its cached payload is damaged.
    // Reject incomplete pointers before adopting saved state or attempting attribution.
    Object contentHash = artifact.opt("contentHash");
    if (!(contentHash instanceof String) || ((String) contentHash).isEmpty()) return false;
    for (String key : new String[] { "appId", "platform", "runtimeVersion" })
      if (!builtin.getString(key).equals(artifact.getString(key))) return false;
    return true;
  }

  private boolean failed(String hash) throws Exception {
    JSONArray failed = state.getJSONArray("failed");
    for (int i = 0; i < failed.length(); i++) if (hash.equals(failed.getString(i))) return true;
    return false;
  }

  private static JSONObject copy(JSONObject value) throws Exception {
    return new JSONObject(value.toString());
  }

  private void persist(JSONObject next) throws Exception {
    trimEvents(next);
    if (storageFailed) throw new Exception("STORAGE_UNAVAILABLE");
    File directory = file.getParentFile();
    Files.createDirectories(directory.toPath());
    File temporary = new File(directory, "." + UUID.randomUUID() + ".tmp");
    try {
      try (FileOutputStream output = new FileOutputStream(temporary)) {
        output.write(next.toString().getBytes(StandardCharsets.UTF_8));
        output.getFD().sync();
      }
      Files.move(
        temporary.toPath(),
        file.toPath(),
        StandardCopyOption.ATOMIC_MOVE,
        StandardCopyOption.REPLACE_EXISTING
      );
      directorySync.sync(directory);
      state = next;
    } catch (Exception error) {
      storageFailed = true;
      throw error;
    } finally {
      Files.deleteIfExists(temporary.toPath());
    }
  }
}
